google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/DynamicFlowControlCallable.java

```diff
@@ -0,0 +1,170 @@
+        new DynamicFlowControlRunnable(
+            flowController,
+            flowControlEvents,
+            dynamicFlowControlStats,
+            targetLatencyMs,
+            adjustingIntervalMs);
+    ApiFuture future = innerCallable.futureCall(request, context);
+    future.addListener(flowControllerRunnable, MoreExecutors.directExecutor());
+    return future;
+  }
+
+  class DynamicFlowControlRunnable implements Runnable {
+    private final FlowController flowController;
+    private final FlowControlEventStats flowControlEvents;
+    private final DynamicFlowControlStats dynamicFlowControlStats;
+    private final long targetLatency;
+    private final long adjustingIntervalMs;
+    private Stopwatch timer;
+
+    private final long slowIncreaseStep;
+    private final long fastIncreaseStep;
+    private final long slowDecreaseStep;
+    private final long fastDecreaseStep;
+    private final double highTargetLatencyMs;
+    private final double lowTargetLatencyMs;
+    private final double highLatencyMs;
+    private final double lowParallelismLatencyMs;
+    private final double lowParallelismLimit;
+    private final long throttlingEventTimeRangeMs;
+
+    DynamicFlowControlRunnable(
+        @Nonnull FlowController flowController,
+        @Nonnull FlowControlEventStats flowControlStats,
+        @Nonnull DynamicFlowControlStats stats,
+        long targetLatency,
+        long adjustingIntervalMs) {
+      this.flowController = flowController;
+      this.flowControlEvents = flowControlStats;
+      this.dynamicFlowControlStats = stats;
+      this.targetLatency = targetLatency;
+      this.adjustingIntervalMs = adjustingIntervalMs;
+      timer = Stopwatch.createStarted();
+
+      // Defining adjusting criteria and adjusting steps
+      // targeting roughly 20% around target latency so there isn't too much churn
+      highTargetLatencyMs = targetLatencyMs * 1.2;
+      lowTargetLatencyMs = targetLatencyMs * 0.8;
+      // when latency is too high, decreasing the thresholds faster
+      highLatencyMs = targetLatencyMs * 3;
+      long maxElementCountLimit = flowController.getMaxElementCountLimit();
+      // make sure the parallelism is not too low
+      lowParallelismLimit = 0.05 * maxElementCountLimit;
+      lowParallelismLatencyMs = targetLatency * 2;
+      // Increase parallelism at a slower rate than decrease. The lower rate should help the system
+      // maintain stability.
+      slowIncreaseStep = Math.round(0.02 * maxElementCountLimit);
+      fastIncreaseStep = Math.round(0.05 * maxElementCountLimit);
+      slowDecreaseStep = Math.round(0.1 * maxElementCountLimit);
+      fastDecreaseStep = Math.round(0.3 * maxElementCountLimit);
+      // only look for throttling events in the past 5 minutes
+      throttlingEventTimeRangeMs = TimeUnit.MINUTES.toMillis(5);
+    }
+
+    @Override
+    public void run() {
+      dynamicFlowControlStats.updateLatency(timer.elapsed(TimeUnit.MILLISECONDS));
```

Comment:
I'm not sure if we want to include all failures here. Maybe just the deadline_exceeded ones?